internal/util/search/search.go

```
@@ -0,0 +1,178 @@
+	Value string
+
+	// ValueRegex is the value regex of the field to be matched
+	ValueRegex string
+
+	// Path is the path of the field to be matched
+	Path string
+
+	// Count is the number of matches
+	Count int
+
+	// Match is the map of file path to list of matched fields
+	Match map[string][]string
+
+	// PutLiteral is the literal to be put at to field
+	// filtered by path and/or value
+	PutLiteral string
+
+	filePath string
+}
+
+// Perform performs the search and replace operation on each node in the package path
+func (sr *SearchReplace) Perform(resourcesPath string) error {
+	inout := &kio.LocalPackageReadWriter{
+		PackagePath:     resourcesPath,
+		NoDeleteFiles:   true,
+		PackageFileName: kptfile.KptFileName,
+	}
+	sr.Match = make(map[string][]string)
+	return kio.Pipeline{
+		Inputs:  []kio.Reader{inout},
+		Filters: []kio.Filter{kio.FilterAll(sr)},
+		Outputs: []kio.Writer{inout},
+	}.Execute()
+}
+
+// Filter parses input node and performs search and replace operation on the node
+func (sr *SearchReplace) Filter(object *yaml.RNode) (*yaml.RNode, error) {
+	filePath, _, err := kioutil.GetFileAnnotations(object)
+	if err != nil {
+		return object, err
+	}
+	sr.filePath = filePath
+
+	if sr.shouldPutLiteralByPath() {
+		return object, sr.putLiteral(object)
+	}
+
+	// traverse the node to perform search/put operation
+	err = accept(sr, object)
+	return object, err
+}
+
+// visitMapping parses mapping node
+func (sr *SearchReplace) visitMapping(object *yaml.RNode, path string) error {
+	return nil
+}
+
+// visitSequence parses sequence node
+func (sr *SearchReplace) visitSequence(object *yaml.RNode, path string) error {
+	// TODO: pmarupaka support sequence nodes
+	return nil
+}
+
+// visitScalar parses scalar node
+func (sr *SearchReplace) visitScalar(object *yaml.RNode, path string) error {
+	regexMatch, err := sr.regexMatch(object.Document().Value)
+	if err != nil {
+		return err
+	}
+
+	pathMatch, err := sr.pathMatch(path)
```

you probably want to do the cheaper operations first. Matching by path is much faster than matching by regex, so you can fail faster by reordering the operations.